### PR TITLE
feat: allow adding cookies from the UI directly and fix the ytdl bug

### PR DESCRIPTION
**Pull request by zdimension**
## Description

Cookies exported from a browser plugin can now be imported directly from the preferences page.

Cookies are now passed to Youtube-DL in a temporary file so writes by Youtube-DL are not saved (this was previously a problem).

### Screenshot

(chrome plugin)
image

(prefs page)
image

## Type of Change
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Dependency update (updates to dependencies)
- [ ] Documentation update (changes to documentation)
- [ ] Repository update (changes to repository files, e.g. `.github/...`)

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated the in code docstring/documentation-blocks for new or existing methods/components

## Branch Updates
LizardByte requires that branches be up-to-date before merging. This means that after any PR is merged, this branch
must be updated before it can be merged. You must also
[Allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I want maintainers to keep my branch updated

### CODE REVIEW

File: Contents/Code/youtube_dl_helper.py [hunk 1]
```diff
@@ -41,74 +44,97 @@ def process_youtube(url):
...
+        )
+
+        if Prefs['str_youtube_cookies']:
+            try:
+                cookies = json.loads(Prefs['str_youtube_cookies'])
+                cookie_jar_file.write('# Netscape HTTP Cookie File\n')
+                for cookie in cookies:
+                    include_subdom = True if cookie['domain'].startswith('.') else False
+                    expiry = int(cookie.get('expiry', 0))
+                    nsbool = lambda b: 'TRUE' if b else 'FALSE'
```
Comment (ReenigneArcher) [inline Contents/Code/youtube_dl_helper.py:66]: The lambda function should probably be moved after this

```py
# get the plugin logger
plugin_logger = logging.getLogger(plugin_identifier)
```

File: Contents/Code/youtube_dl_helper.py [hunk 2]
```diff
@@ -41,74 +44,97 @@ def process_youtube(url):
...
+                    values = [
+                        cookie['domain'],
+                        nsbool(include_subdom),
+                        cookie['path'],
+                        nsbool(cookie['secure']),
+                        str(expiry),
+                        cookie['name'],
+                        cookie['value']
+                    ]
+                    cookie_jar_file.write('\t'.join(values) + '\n')
```
Comment (ReenigneArcher) [inline Contents/Code/youtube_dl_helper.py:76]:
```suggestion
                    cookie_jar_file.write('{}\n'.format('\t'.join(values)))
```

I'm not really a fan of string concatenation.

File: Contents/Code/youtube_dl_helper.py [hunk 3]
```diff
@@ -41,74 +44,97 @@ def process_youtube(url):
...
+            except Exception as exc:
+                Log.Exception('Failed to write YouTube cookies to file, will try anyway. Error: %s' % exc)
```
Comment (ReenigneArcher) [inline Contents/Code/youtube_dl_helper.py:78]:
```suggestion
            except Exception as e:
                Log.Exception('Failed to write YouTube cookies to file, will try anyway. Error: {}'.format(e))
```

I'm slowly working towards replacing all `%s` with `.format()` when possible.

File: docs/source/about/troubleshooting.rst
Comment (ReenigneArcher) [inline docs/source/about/troubleshooting.rst:1]: All settings should also be documented in `about/usage.rst`.

**Review by ReenigneArcher — CHANGES_REQUESTED** [review]

File: docs/source/about/usage.rst [hunk 4]
```diff
@@ -230,6 +230,16 @@ Description
 Default
    None
 
+YouTube Cookies
+^^^^^^^^^^^^^^^^
+
+Description
+   The cookies to use for the requests to YouTube. Should be in Chromium JSON export format. 
+   [Example exporter](https://chrome.google.com/webstore/detail/get-cookiestxt/bgaddhkoddajcdgocldbbfleckgcbcid).
```
Comment (ReenigneArcher) [inline docs/source/about/usage.rst:238]:
```suggestion
   `Example exporter <https://chrome.google.com/webstore/detail/get-cookiestxt/bgaddhkoddajcdgocldbbfleckgcbcid>`__.
```

File: Contents/Code/youtube_dl_helper.py [hunk 5]
```diff
@@ -13,13 +15,30 @@
...
-from constants import plugin_identifier, cookie_jar_file
+from constants import plugin_identifier
 from typing import Optional
 import youtube_dl
 
 # get the plugin logger
 plugin_logger = logging.getLogger(plugin_identifier)
 
+def nsbool(value):
+    # type: (bool) -> str
```
Comment (ReenigneArcher) [inline Contents/Code/youtube_dl_helper.py:26]:
```suggestion
# get the plugin logger
plugin_logger = logging.getLogger(plugin_identifier)


def nsbool(value):
    # type: (bool) -> str
```

flake8 wants 2 blank lines before functions

**Review by ReenigneArcher — CHANGES_REQUESTED** [review]

File: docs/source/about/usage.rst (hunk 4 above)
Comment (zdimension) [reply docs/source/about/usage.rst:238]: 🤦

**Review by ReenigneArcher — APPROVED** [review]

**Pull request merged**